feat: implement enterprise search agent tool and related functionality

m-baden opened the pull request:
### Link to Issue or Description of Change

**1. Link to an existing issue (if applicable):**

- Closes: #3412

### Testing Plan

**Unit Tests:**

- [x] I have added or updated unit tests for my change.
- [x] All unit tests pass locally.

A new test file `tests/unittests/tools/test_enterprise_search_agent_tool.py` has been added to test the new functionality. The tests are passing.

All unit tests were run. There are some collection errors that are unrelated to this change.

```
$ pytest tests/unittests
================================================= 3182 passed, 2410 warnings in 35.24s ==================================================
```

**Manual End-to-End (E2E) Tests:**

Manual E2E tests were performed using the following agent definition:

```python
from google.adk.agents.llm_agent import Agent
from google.adk.tools.enterprise_search_tool import EnterpriseWebSearchTool
from google.adk.tools.google_search_tool import GoogleSearchTool


root_agent = Agent(
    model='gemini-2.5-flash',
    name='root_agent',
    description="A helpful assistant that can use tools",
    instruction="You are a helpful assistant. Always use both search tools to answer user queries.",
    tools=[EnterpriseWebSearchTool(
        bypass_multi_tools_limit=True), GoogleSearchTool(bypass_multi_tools_limit=True)],
)
```
<img width="1728" height="995" alt="image" src="https://github.com/user-attachments/assets/bee041da-9022-4512-8ea2-baa209c30888" />


Manual E2E tests were also performed with an agent using a custom tool alongside the Google Search Tool:

```python
from google.adk.agents.llm_agent import Agent
from google.adk.tools.enterprise_search_tool import EnterpriseWebSearchTool

def get_current_time(city: str) -> dict:
    """Returns the current time in a specified city."""
    return {"status": "success", "city": city, "time": "10:30 AM"}

root_agent = Agent(
    model='gemini-2.5-flash',
    name='root_agent',
    description="A helpful assistant that can use tools",
    instruction="You are a helpful assistant. Always use both search tools to answer user queries.",
    tools=[get_current_time, GoogleSearchTool(bypass_multi_tools_limit=True)],
)
```

These tests confirmed that the `GoogleSearchTool` can be used alongside other custom tools when `bypass_multi_tools_limit=True` is set.
<img width="1728" height="992" alt="image" src="https://github.com/user-attachments/assets/62b72f1f-f86c-4b32-9b4b-1eee177e2bdb" />

The agent was run with various queries to ensure that both `EnterpriseWebSearchTool` and `GoogleSearchTool` are correctly invoked and their results are processed as expected. The `bypass_multi_tools_limit=True` parameter was verified to allow both tools to be used simultaneously.

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) document.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have manually tested my changes end-to-end.
- [x] Any dependent changes have been merged and published in downstream modules.

### Additional context

This PR is similar to https://github.com/google/adk-python/commit/d3148dacc97f0a9a39b6d7a9640f7b7b0d6f9a6c

m-baden commented on src/google/adk/flows/llm_flows/base_llm_flow.py:
```diff
@@ -846,7 +846,11 @@ async def _maybe_add_grounding_metadata(
+          tool.name == 'google_search_agent'
+          or tool.name == 'enterprise_search_agent'
+          for tool in tools
+      ):
```
done

m-baden commented on tests/unittests/tools/test_enterprise_search_agent_tool.py:
```diff
@@ -0,0 +1,94 @@
+
+  @mark.asyncio
+  @mock.patch('google.adk.tools._search_agent_tool._SearchAgentTool.run_async')
+  async def test_run_async_succeeds(self, mock_run_async):
```
done

m-baden commented on src/google/adk/tools/enterprise_search_agent_tool.py:
```diff
@@ -0,0 +1,55 @@
+
+  Attributes:
+    model: The model to use for the sub-agent.
+  """
```
done

ryanaiagent commented:
Hi @m-baden , Thank you for your work on this pull request. We appreciate the effort you've invested. 
Can you please fix lint errors. You can run autoformat.sh

m-baden commented:
Hi @ryanaiagent, thanks for taking the time to look into it. The lint errors are fixed now.

ryanaiagent commented:
Hi @m-baden , Thank you for your patience here. I apologize for the delay in getting to this review; I know this has been sitting for a while. This PR has merge conflicts that require changes from your end. Could you please rebase your branch with the latest main branch to address these? Once this is complete, please let us know so we can proceed with the review.

m-baden committed:
feat: implement enterprise search agent tool and related updates

- Refactor search agent tools to inherit from a common base class

- Streamline tool handling in LLM agent with dedicated workarounds

- Enhance state filtering to exclude temporary states

m-baden commented:
@GWeale @rohityan This PR is almost open for half a year. When can I expect a review?
